### ISSUE TITLE
[gmp:x64-windows] build failure

### ISSUE THREAD

**Issue by DanielZint** [microsoft/vcpkg#17286]
**Host Environment**
- OS: Windows Version 20H2 (OS Build 19042.867)
- Compiler: Visual Studio 2019 (v142)

**To Reproduce**
`./vcpkg install yasm-tool:x86-windows`
`./vcpkg install gmp:x64-windows`

**Failure logs** 
- powershell output:
```
PS C:\Program Files\vcpkg> .\vcpkg.exe install gmp:x64-windows
Computing installation plan...
The following packages will be built and installed:
    gmp[core]:x64-windows -> 6.2.1#4
  * vs-yasm[core]:x64-windows -> 0.5.0#1
  * yasm-tool-helper[core]:x64-windows -> 2020-10-12
Additional packages (*) will be modified to complete this operation.
Detecting compiler hash for triplet x64-windows...
Using cached binary package: C:\Users\da_zi\AppData\Local\vcpkg\archives\fd\fd97b8014eec5ec67d82db379ddaa2d5efba3ae3.zip
Using cached binary package: C:\Users\da_zi\AppData\Local\vcpkg\archives\37\3756e0a9bc33894f63c876705fe156c735540998.zip
Could not locate cached archive: C:\Users\da_zi\AppData\Local\vcpkg\archives\16\16e6ce3a3379ecd147970f50417332164cab3ed3.zip
Starting package 1/3: vs-yasm:x64-windows
Building package vs-yasm[core]:x64-windows...
Building package vs-yasm[core]:x64-windows... done
Installing package vs-yasm[core]:x64-windows...
Installing package vs-yasm[core]:x64-windows... done
Elapsed time for package vs-yasm:x64-windows: 23.94 ms
Starting package 2/3: yasm-tool-helper:x64-windows
Building package yasm-tool-helper[core]:x64-windows...
Building package yasm-tool-helper[core]:x64-windows... done
Installing package yasm-tool-helper[core]:x64-windows...
Installing package yasm-tool-helper[core]:x64-windows... done
Elapsed time for package yasm-tool-helper:x64-windows: 13.47 ms
Starting package 3/3: gmp:x64-windows
Building package gmp[core]:x64-windows...
-- Using cached C:/Program Files/vcpkg/downloads/ShiftMediaProject-gmp-0018c44e8dfcc3b64b43e0aea4b3f419f0b65fd0.tar.gz
-- Extracting source C:/Program Files/vcpkg/downloads/ShiftMediaProject-gmp-0018c44e8dfcc3b64b43e0aea4b3f419f0b65fd0.tar.gz
-- Applying patch vs.build.patch
-- Applying patch runtime.patch
-- Applying patch adddef.patch
-- Using source at C:/Program Files/vcpkg/buildtrees/gmp/src/19f0b65fd0-9b395bc164.clean
-- Building SMP/libgmp.sln for Release
CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:105 (message):
    Command failed: msbuild "C:/Program Files/vcpkg/buildtrees/gmp/x64-windows-rel/19f0b65fd0-9b395bc164.clean/SMP/libgmp.sln" /p:Configuration=ReleaseDLL "/p:YasmPath=C:\\Program Files\\vcpkg\\installed\\x86-windows\\tools\\yasm-tool\\yasm.exe" /t:Rebuild /p:Platform=x64 /p:PlatformToolset=v142 /p:VCPkgLocalAppDataDisabled=true /p:UseIntelMKL=No /p:WindowsTargetPlatformVersion=10.0.19041.0 /p:VcpkgTriplet=x64-windows "/p:VcpkgInstalledDir=C:/Program Files/vcpkg/installed" /p:VcpkgManifestInstall=false /m "/p:ForceImportBeforeCppTargets=C:/Program Files/vcpkg/scripts/buildsystems/msbuild/vcpkg.targets" /p:VcpkgApplocalDeps=false
    Working Directory: C:/Program Files/vcpkg/buildtrees/gmp/x64-windows-rel/19f0b65fd0-9b395bc164.clean
    Error code: 1
    See logs for more information:
      C:\Program Files\vcpkg\buildtrees\gmp\build-x64-windows-rel-out.log

Call Stack (most recent call first):
  scripts/cmake/vcpkg_install_msbuild.cmake:165 (vcpkg_execute_required_process)
  ports/gmp/portfile.cmake:53 (vcpkg_install_msbuild)
  scripts/ports.cmake:142 (include)


Error: Building package gmp:x64-windows failed with: BUILD_FAILED
Please ensure you're using the latest portfiles with `.\vcpkg update`, then
submit an issue at https://github.com/Microsoft/vcpkg/issues including:
  Package: gmp:x64-windows
  Vcpkg version: 2021-01-13-d67989bce1043b98092ac45996a8230a059a2d7e

Additionally, attach any relevant sections from the log files above.
```

-Please attach any additional failure logs mentioned in the console output.
[build-x64-windows-rel-out.log](https://github.com/microsoft/vcpkg/files/6313497/build-x64-windows-rel-out.log)

**Additional context**
I am aware that yasm-tool:x86-windows needs to be installed first but that did not solve this issue for me.

**Comment by NancyLi1013** [comment]
Hi @DanielZint 

Thanks for posting this issue.

```
    The system cannot find the path specified.
     2>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\VC\v160\Microsoft.CppCommon.targets(285,5): error MSB3073: The command "if exist "C:\Program Files\vcpkg\buildtrees\gmp\x64-windows-rel\19f0b65fd0-9b395bc164.clean\SMP\..\msvc\"\include\gmp.h ( [C:\Program Files\vcpkg\buildtrees\gmp\x64-windows-rel\19f0b65fd0-9b395bc164.clean\SMP\libgmp.vcxproj]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\VC\v160\Microsoft.CppCommon.targets(285,5): error MSB3073: del /f /q "C:\Program Files\vcpkg\buildtrees\gmp\x64-windows-rel\19f0b65fd0-9b395bc164.clean\SMP\..\msvc\"\include\gmp.h [C:\Program Files\vcpkg\buildtrees\gmp\x64-windows-rel\19f0b65fd0-9b395bc164.clean\SMP\libgmp.vcxproj]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\VC\v160\Microsoft.CppCommon.targets(285,5): error MSB3073: ) [C:\Program Files\vcpkg\buildtrees\gmp\x64-windows-rel\19f0b65fd0-9b395bc164.clean\SMP\libgmp.vcxproj]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\VC\v160\Microsoft.CppCommon.targets(285,5): error MSB3073: if exist C:\Program Files\vcpkg\buildtrees\gmp\x64-windows-rel\19f0b65fd0-9b395bc164.clean\SMP\..\msvc\\licenses\gmp.txt ( [C:\Program Files\vcpkg\buildtrees\gmp\x64-windows-rel\19f0b65fd0-9b395bc164.clean\SMP\libgmp.vcxproj]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\VC\v160\Microsoft.CppCommon.targets(285,5): error MSB3073: del /f /q C:\Program Files\vcpkg\buildtrees\gmp\x64-windows-rel\19f0b65fd0-9b395bc164.clean\SMP\..\msvc\\licenses\gmp.txt [C:\Program Files\vcpkg\buildtrees\gmp\x64-windows-rel\19f0b65fd0-9b395bc164.clean\SMP\libgmp.vcxproj]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\VC\v160\Microsoft.CppCommon.targets(285,5): error MSB3073: ) [C:\Program Files\vcpkg\buildtrees\gmp\x64-windows-rel\19f0b65fd0-9b395bc164.clean\SMP\libgmp.vcxproj]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\VC\v160\Microsoft.CppCommon.targets(285,5): error MSB3073: :VCEnd" exited with code 1. [C:\Program Files\vcpkg\buildtrees\gmp\x64-windows-rel\19f0b65fd0-9b395bc164.clean\SMP\libgmp.vcxproj]
     2>Done Building Project "C:\Program Files\vcpkg\buildtrees\gmp\x64-windows-rel\19f0b65fd0-9b395bc164.clean\SMP\libgmp.vcxproj" (Rebuild target(s)) -- FAILED.
     1>Done Building Project "C:\Program Files\vcpkg\buildtrees\gmp\x64-windows-rel\19f0b65fd0-9b395bc164.clean\SMP\libgmp.sln" (Rebuild target(s)) -- FAILED.
```
Could you please check the path in the command line?

**Comment by O-Tom** [comment]
Here is an excerpt from the buildlog that @DanielZint provided:
`"C:\Program Files\vcpkg\installed\x86-windows\tools\yasm-tool\yasm.exe" -Xvc -f win64 -o "C:\Program Files\vcpkg\buildtrees\gmp\x64-windows-rel\19f0b65fd0-9b395bc164.clean\SMP\obj\ReleaseDLL\x64\libgmp\addlsh1_n.obj" -rraw -pgas  "C:\Program Files\vcpkg\buildtrees\gmp\x64-windows-rel\19f0b65fd0-9b395bc164.clean\SMP\mpn\x86_64\addlsh1_n.s"`

He also wrote that to compile gmp for the `x64-windows` triplet he used the following commands:
```
./vcpkg install yasm-tool:x86-windows
./vcpkg install gmp:x64-windows
```

How comes a x64 target is requiring an x86 tool (yasm) to be installed ?
Could the `gmp` port be modified to use the x64 version of `yasm-tool` or declare a dependency on the x86 version if it really needs it ?

**Comment by DanielZint** [comment]
I think I found the issue. When installing vcpkg in a different folder, then everything works as expected:
`C:\Program Files\vcpkg` - does not work
`C:\libs\vcpkg` - works
I guess it has something to do with admin rights. The folder 
`C:\Program Files\vcpkg\buildtrees\gmp\x64-windows-rel\19f0b65fd0-9b395bc164.clean\msvc\`
is empty, whereas
`C:\libs\vcpkg\buildtrees\gmp\x64-windows-rel\19f0b65fd0-9b395bc164.clean\msvc\`
is not.

It is just confusing for me because I used the *Program Files* folder already before and never had an issue. I am not sure though if maybe some Windows update messed that up.

Basically my problem was solved now but I have one question remaining. When using GMP I got the error message that gmp.dll was not found. I copied it manually from `C:\libs\vcpkg\installex\x64-windows\bin` to my executable folder and then everything worked fine. This issue might also be caused by cgal, the library that I actually want to use. Should I open a new issue for that or am I just missing something obvious?

**Comment by basselin7u** [comment]
Hello, 
I had the same issue during the installation. 
It is not a problem of writing rights, but a problem of space in the folder name.

**Comment by DanielZint** [comment]
Thank you for pointing that out @basselin7u !

**Comment by NancyLi1013** [comment]
@DanielZint 

Thanks for your further investigation.
For the original issue, it seems to be clear that this is caused by space in path.

For the new issue you mentioned, you can open a new issue.
